sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala

```diff
@@ -91,6 +107,37 @@ class DateFunctionsSuite extends QueryTest with SharedSQLContext {
       Row(Date.valueOf("2015-01-01")) :: Nil)
   }
 
+  test("date comparison with date strings with session local timezone") {
+    val df = Seq(
+      (1, Date.valueOf("2015-12-31")),
+      (2, Date.valueOf("2016-01-01"))).toDF("i", "t")
+
+    checkAnswer(
+      df.select("t").filter($"t" <= "2016-01-01"),
+      Seq(
+        Row(Date.valueOf("2015-12-31")),
+        Row(Date.valueOf("2016-01-01"))))
+
+    checkAnswer(
+      df.select("t").filter($"t" >= "2016"),
+      Seq(
+        Row(Date.valueOf("2016-01-01"))))
+
+    withSQLConf(SQLConf.SESSION_LOCAL_TIMEZONE.key -> "GMT") {
+
+      checkAnswer(
+        df.select("t").filter($"t" <= "2016-01-01"),
+        Seq(
+          Row(Date.valueOf("2015-12-31")),
+          Row(Date.valueOf("2016-01-01"))))
+
+      checkAnswer(
+        df.select("t").filter($"t" >= "2016"),
+        Seq(
+          Row(Date.valueOf("2016-01-01"))))
```

Comment:
nit: we can just write `Row(...)` here, `checkAnswer` has an overload version which takes a single `Row`